Objects/dictobject.c

```diff
@@ -2591,6 +2581,102 @@ dict_merge(PyObject *a, PyObject *b, int override)
     return 0;
 }
 
+/* Implementation of d.update(obj). There are two cases:
+ * - if obj has a "keys" method, assume that it is a mapping
+ * - otherwise, obj must be an iterable of (key, value) pairs
+ */
+static int
+dict_merge_from_any(PyObject *self, PyObject *obj)
+{
+    /* Fast path for dict subclasses */
+    if (PyDict_Check(obj)) {
+        return dict_merge(self, obj, 1);
+    }
+
+    _Py_IDENTIFIER(keys);
+    PyObject *func;
+    if (_PyObject_LookupAttrId(obj, &PyId_keys, &func) < 0) {
+        return -1;
+    }
+    if (func != NULL) {
+        Py_DECREF(func);
+        return dict_merge(self, obj, 1);
+    }
+    else {
+        return PyDict_MergeFromSeq2(self, obj, 1);
+    }
+}
+
+/* Insert the key:value pairs from the arrays keys and values (of length n).
+ * If a key is not a string, raise TypeError */
+static int
+dict_merge_from_kw_arrays(PyDictObject *mp, PyObject *const *keys, PyObject *const *values, Py_ssize_t n)
+{
+    /* Do one big resize at the start, rather than
+     * incrementally resizing as we insert new items.  Expect
+     * that there will be no (or few) overlapping keys.
+     */
+    if (USABLE_FRACTION(mp->ma_keys->dk_size) < n) {
+        if (dictresize(mp, ESTIMATE_SIZE(mp->ma_used + n))) {
+           return -1;
+        }
+    }
+
+    for (Py_ssize_t i = 0; i < n; i++) {
+        PyObject *key = keys[i];
+        if (!PyUnicode_Check(key)) {
+            PyErr_SetString(PyExc_TypeError, "keywords must be strings");
```

Comment:
Do we really have to check here that the FASTCALL protocol is used properly? Maybe use an assertion instead? What do you think @methane?

Reply:
> Do we really have to check here that the FASTCALL protocol is used properly?

This is currently the only place where this is checked. If we remove this check here, then we must check it somewhere else (unless you are suggesting that `dict.update(**d)` should be allowed even if `d` has non-string keys).

There is a test for this in `test_dict.py`:
```python
    def test_invalid_keyword_arguments(self):
        class Custom(dict):
            pass
        for invalid in {1 : 2}, Custom({1 : 2}):
            with self.assertRaises(TypeError):
                dict(**invalid)
            with self.assertRaises(TypeError):
                {}.update(**invalid)
```

Reply:
For METH_VARARGS and METH_VARARGS|METH_KEYWORDS, we don't check arguments type at runtime in release mode. But there are plenty of assertions. Example:

```
PyObject *
PyObject_Call(PyObject *callable, PyObject *args, PyObject *kwargs)
{
    assert(!PyErr_Occurred());
    assert(PyTuple_Check(args));
    assert(kwargs == NULL || PyDict_Check(kwargs));
```

METH_VARARGS|METH_KEYWORDS should only get strings for kwargs keys, but that isn't checked neither.

If you misuse the Python C API, you are on your own: Python is likely to crash, and I consider that it's the expected behavior.

Reply:
I don't understand what you're trying to suggest here. Simply removing the check is not an option, as it causes testsuite failures.

Reply:
Interestingly, `OrderedDict` does allow non-string keys:
```
>>> from collections import OrderedDict
>>> OrderedDict(**{1:2})
OrderedDict([(1, 2)])
```
This is probably a bug, but it shows that nothing in the core interpreter prevents non-string keys.

Reply:
I just posted to python-dev about this issue.

Reply:
Fixed by #14682